File: vkxml2java/codegen/ctype.py
from dataclasses import dataclass

from ..entity import Registry
from ..ident import Identifier
from ..vktype import Type, IdentifierType, ArrayType, PointerType


@dataclass
class CType:
    def java_type(self) -> str:
        raise NotImplementedError()

    def java_layout(self) -> str:
        raise NotImplementedError()

    def java_layout_type(self) -> str:
        raise NotImplementedError()

    def c_type(self) -> str:
        raise NotImplementedError()


@dataclass
class CPointerType(CType):
    pointee: CType
    const: bool
    comment: str | None = None

    def java_type(self) -> str:
        if self.comment is not None:
            return f'@pointer(comment="{self.comment}") MemorySegment'
        elif self.pointee == CTYPE_VOID:
            return '@pointer(comment="void*") MemorySegment'

        return 'MemorySegment'

    def java_layout(self) -> str:
        if self.pointee == CTYPE_VOID:
            return 'ValueLayout.ADDRESS'
        else:
            return 'ValueLayout.ADDRESS.withTargetLayout(' + self.pointee.java_layout() + ')'

    def java_layout_type(self) -> str:
        return 'AddressLayout'

    def c_type(self) -> str:
        return f'{self.pointee.c_type()} {'const' if self.const else ''}*'


@dataclass
class CHandleType(CType):
    name: Identifier

    def java_type(self) -> str:
        return f'@pointer(target={self.name}.class) MemorySegment'

    def java_layout(self) -> str:
        return 'ValueLayout.ADDRESS'

    def java_layout_type(self) -> str:
        return 'AddressLayout'

    def c_type(self) -> str:
        return f'{self.name}'


@dataclass
class CEnumType(CType):
    name: str
    bitwidth: int | None = None

    def java_type(self) -> str:
        name = self.non_flagbits_type_name()
        if self.bitwidth is None or self.bitwidth == 32:
            return f'@enumtype({name}.class) int'
        elif self.bitwidth == 64:
            return f'@enumtype({name}.class) long'
        else:
            raise Exception(f'unsupported bitwidth: {self.bitwidth}')

    def java_layout(self) -> str:
        if self.bitwidth is None or self.bitwidth == 32:
            return 'ValueLayout.JAVA_INT'
        elif self.bitwidth == 64:
            return 'ValueLayout.JAVA_LONG'
        else:
            raise Exception(f'unsupported bitwidth: {self.bitwidth}')

    def java_layout_type(self) -> str:
        if self.bitwidth is None or self.bitwidth == 32:
            return 'OfInt'
        elif self.bitwidth == 64:
            return 'OfLong'
        else:
            raise Exception(f'unsupported bitwidth: {self.bitwidth}')

    def c_type(self) -> str:
        return f'enum {self.name}'

    def non_flagbits_type_name(self) -> str:
        return self.name.replace('FlagBits', 'Flags')


@dataclass
class CArrayType(CType):
    element: CType
    size: int | str

    def java_type(self) -> str:
        return f'{self.element.java_type()}[]'

    def java_layout(self) -> str:
        return f'MemoryLayout.sequenceLayout({self.size}, {self.element.java_layout()})'

    def java_layout_type(self) -> str:
        return 'SequenceLayout'

    def c_type(self) -> str:
        return f'{self.element.c_type()}[{self.size}]'


class CNonRefType(CType):
    def vk4j_array_type(self) -> str:
        raise NotImplementedError()

    def vk4j_array_type_no_sign(self) -> str:
        raise NotImplementedError()

    def vk4j_ptr_type(self) -> str:
        raise NotImplementedError()

    def vk4j_ptr_type_no_sign(self) -> str:
        raise NotImplementedError()


@dataclass
class CFixedIntType(CNonRefType):
    c_name: str
    byte_size: int
    unsigned: bool

    def java_type(self) -> str:
        unsigned_prefix = '@unsigned ' if self.unsigned else ''
        if self.c_name == 'char' or self.byte_size == 1:
            return f'{unsigned_prefix}byte'
        elif self.byte_size == 2:
            return f'{unsigned_prefix}short'
        elif self.byte_size == 4:
            return f'{unsigned_prefix}int'
        elif self.byte_size == 8:
            return f'{unsigned_prefix}long'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def java_layout(self) -> str:
        if self.c_name == 'char' or self.byte_size == 1:
            return 'ValueLayout.JAVA_BYTE'
        elif self.byte_size == 2:
            return 'ValueLayout.JAVA_SHORT'
        elif self.byte_size == 4:
            return 'ValueLayout.JAVA_INT'
        elif self.byte_size == 8:
            return 'ValueLayout.JAVA_LONG'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def java_layout_type(self) -> str:
        if self.c_name == 'char' or self.byte_size == 1:
            return 'OfByte'
        elif self.byte_size == 2:
            return 'OfShort'
        elif self.byte_size == 4:
            return 'OfInt'
        elif self.byte_size == 8:
            return 'OfLong'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def c_type(self) -> str:
        return self.c_name

    def vk4j_array_type(self) -> str:
        unsigned_prefix = '@unsigned ' if self.unsigned else ''
        return f'{unsigned_prefix}{self.vk4j_array_type_nosign()}'

    def vk4j_array_type_nosign(self) -> str:
        if self.c_name == 'char' or self.byte_size == 1:
            return 'ByteArray'
        elif self.byte_size == 2:
            return 'DoubleArray'
        elif self.byte_size == 4:
            return 'IntArray'
        elif self.byte_size == 8:
            return 'LongArray'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def vk4j_ptr_type(self) -> str:
        unsigned_prefix = '@unsigned ' if self.unsigned else ''
        return f'{unsigned_prefix}{self.vk4j_ptr_type_no_sign()}'

    def vk4j_ptr_type_no_sign(self) -> str:
        if self.c_name == 'char' or self.byte_size == 1:
            return f'BytePtr'
        elif self.byte_size == 2:
            return f'ShortPtr'
        elif self.byte_size == 4:
            return f'IntPtr'
        elif self.byte_size == 8:
            return f'LongPtr'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')


@dataclass
class CPlatformDependentIntType(CNonRefType):
    c_name: str
    array_name: str
    ptr_name: str
    java_layout_: str
    java_type_: str

    def java_type(self) -> str:
        return self.java_type_

    def java_layout(self) -> str:
        return self.java_layout_

    def java_layout_type(self) -> str:
        raise Exception(f'should not call `java_layout_type` on platform dependent int')

    def c_type(self) -> str:
        return self.c_name

    def vk4j_array_type(self) -> str:
        return self.array_name

    def vk4j_array_type_no_sign(self) -> str:
        return self.array_name

    def vk4j_ptr_type(self) -> str:
        return self.ptr_name

    def vk4j_ptr_type_no_sign(self) -> str:
        return self.ptr_name


@dataclass
class CFloatType(CNonRefType):
    byte_size: int

    def java_type(self) -> str:
        if self.byte_size == 4:
            return 'float'
        elif self.byte_size == 8:
            return 'double'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def java_layout(self) -> str:
        if self.byte_size == 4:
            return 'ValueLayout.JAVA_FLOAT'
        elif self.byte_size == 8:
            return 'ValueLayout.JAVA_DOUBLE'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def java_layout_type(self) -> str:
        if self.byte_size == 4:
            return 'OfFloat'
        elif self.byte_size == 8:
            return 'OfDouble'
        else:
            raise Exception(f'unsupported byte size: {self.byte_size}')

    def c_type(self) -> str:
        return 'float' if self.byte_size == 4 else 'double'

    def vk4j_array_type(self) -> str:
        return 'FloatArray' if self.byte_size == 4 else 'DoubleArray'

    def vk4j_array_type_no_sign(self) -> str:
        return 'FloatArray' if self.byte_size == 4 else 'DoubleArray'

    def vk4j_ptr_type(self) -> str:
        return 'FloatPtr' if self.byte_size == 4 else 'DoublePtr'

    def vk4j_ptr_type_no_sign(self) -> str:
        return 'FloatPtr' if self.byte_size == 4 else 'DoublePtr'


@dataclass
class CBoolType(CType):
    def java_type(self) -> str:
        return 'boolean'

    def java_layout(self) -> str:
        return 'ValueLayout.JAVA_BOOLEAN'

    def java_layout_type(self) -> str:
        return 'OfBoolean'

    def c_type(self) -> str:
        return 'bool'


@dataclass
class CStructType(CType):
    name: str

    def java_type(self) -> str:
        return self.name

    def java_layout(self) -> str:
        return f'{self.name}.LAYOUT'

    def java_layout_type(self) -> str:
        return 'StructLayout'

    def c_type(self) -> str:
        return f'struct {self.name}'


@dataclass
class CUnionType(CType):
    name: str

    def java_type(self) -> str:
        return self.name

    def java_layout(self) -> str:
        return f'{self.name}.LAYOUT'

    def java_layout_type(self) -> str:
        return 'UnionLayout'

    def c_type(self) -> str:
        return f'union {self.name}'


@dataclass
class CVoidType(CType):
    pass

    def java_type(self) -> str:
        return 'void'

    def java_layout(self) -> str:
        raise Exception(f'should not call `java_layout` on void')

    def java_layout_type(self) -> str:
        raise Exception(f'should not call `java_layout_type` on void')

    def c_type(self) -> str:
        return 'void'


CTYPE_INT8: CFixedIntType = CFixedIntType('int8_t', 1, False)
CTYPE_INT16: CFixedIntType = CFixedIntType('int16_t', 2, False)
CTYPE_INT32: CFixedIntType = CFixedIntType('int32_t', 4, False)
CTYPE_INT64: CFixedIntType = CFixedIntType('int64_t', 8, False)
CTYPE_UINT8: CFixedIntType = CFixedIntType('uint8_t', 1, True)
CTYPE_UINT16: CFixedIntType = CFixedIntType('uint16_t', 2, True)
CTYPE_UINT32: CFixedIntType = CFixedIntType('uint32_t', 4, True)
CTYPE_UINT64: CFixedIntType = CFixedIntType('uint64_t', 8, True)
CTYPE_FLOAT: CType = CFloatType(4)
CTYPE_DOUBLE: CType = CFloatType(8)

CTYPE_INT: CType = CTYPE_INT32
CTYPE_LONG: CType = CPlatformDependentIntType('long', 'CLongArray', 'CLongPtr', 'NativeLayout.C_LONG', 'long')
CTYPE_SIZET: CType = CPlatformDependentIntType('size_t', 'CSizeTArray', 'CSizeTPtr', 'NativeLayout.C_SIZE_T', '@unsigned long')

CTYPE_VOID: CType = CVoidType()
CTYPE_PVOID: CType = CPointerType(CTYPE_VOID, False)


KNOWN_TYPES: dict[str, CType] = {
    # Fundatmental types
    'char': CTYPE_INT8,

    'int8_t': CTYPE_INT8,
    'int16_t': CTYPE_INT16,
    'int32_t': CTYPE_INT32,
    'int64_t': CTYPE_INT64,
    'uint8_t': CTYPE_UINT8,
    'uint16_t': CTYPE_UINT16,
    'uint32_t': CTYPE_UINT32,
    'uint64_t': CTYPE_UINT64,
    'float': CTYPE_FLOAT,
    'double': CTYPE_DOUBLE,
    'bool': CBoolType(),
    'void': CTYPE_VOID,
    'int': CTYPE_INT,
    'long': CTYPE_LONG,
    'size_t': CTYPE_SIZET,

    # Vulkan base types
    'VkSampleMask': CTYPE_UINT32,
    'VkBool32': CTYPE_UINT32,
    'VkFlags': CTYPE_UINT32,
    'VkFlags64': CTYPE_UINT64,
    'VkDeviceSize': CTYPE_UINT64,
    'VkDeviceAddress': CTYPE_UINT64,
    'VkRemoteAddressNV': CTYPE_PVOID,

    # Android
    'ANativeWindow': CTYPE_VOID,
    'AHardwareBuffer': CTYPE_VOID,

    # DirectFB
    'IDirectFB': CTYPE_VOID,
    'IDirectFBSurface': CTYPE_VOID,

    # iOS / MacOS
    'CAMetalLayer': CTYPE_VOID,
    'GgpFrameToken': CTYPE_UINT32,
    'GgpStreamDescriptor': CTYPE_UINT32,
    'IOSurfaceRef': CTYPE_PVOID,
    'MTLBuffer_id': CTYPE_PVOID,
    'MTLCommandQueue_id': CTYPE_PVOID,
    'MTLDevice_id': CTYPE_PVOID,
    'MTLSharedEvent_id': CTYPE_PVOID,
    'MTLTexture_id': CTYPE_PVOID,

    # QNX
    '_screen_buffer': CTYPE_VOID,
    '_screen_context': CTYPE_VOID,
    '_screen_window': CTYPE_VOID,

    # Wayland
    'wl_display': CTYPE_VOID,
    'wl_surface': CTYPE_VOID,

    # Windows
    'DWORD': CTYPE_UINT32,
    'HANDLE': CTYPE_PVOID,
    'HINSTANCE': CTYPE_PVOID,
    'HMONITOR': CTYPE_PVOID,
    'HWND': CTYPE_PVOID,
    'LPCWSTR': CPointerType(CTYPE_UINT16, True),
    'SECURITY_ATTRIBUTES': CTYPE_VOID,

    # X11
    'Display': CTYPE_PVOID,
    'RROutput': CTYPE_LONG,
    'VisualID': CTYPE_LONG,
    'Window': CTYPE_LONG,
    'xcb_connection_t': CTYPE_VOID,
    'xcb_visualid_t': CTYPE_UINT32,
    'xcb_window_t': CTYPE_UINT32,
    'zx_handle_t': CTYPE_UINT32,

    # NvSciBuf / NvSciSync
    'NvSciBufAttrList': CTYPE_PVOID,
    'NvSciBufObj': CTYPE_PVOID,
    'NvSciSyncAttrList': CTYPE_PVOID,
    'NvSciSyncObj': CTYPE_PVOID,
    'NvSciSyncFence': CArrayType(CTYPE_UINT64, 6),
}


def lower_type(registry: Registry, type_: Type) -> CType:
    if isinstance(type_, IdentifierType):
        return lower_identifier_type(registry, type_)
    elif isinstance(type_, ArrayType):
        if not type_.length.value.isnumeric():
            if type_.length not in registry.constants:
                raise Exception(f'array typed referred to an unknown constant: {type_.length.value}')
        element_type = lower_type(registry, type_.element)
        return CArrayType(element_type, type_.length.value)
    elif isinstance(type_, PointerType):
        pointee = lower_type(registry, type_.pointee)
        return CPointerType(pointee, type_.const)


def lower_identifier_type(registry: Registry, ident_type: IdentifierType) -> CType:
    ident = ident_type.identifier
    ident_value = ident.value

    if ident_value in KNOWN_TYPES:
        ret = KNOWN_TYPES[ident_value]
        return ret
    elif ident in registry.enums:
        return CEnumType(ident.value)
    elif ident in registry.bitmasks:
        bitmask = registry.bitmasks[ident]
        return CEnumType(ident.value, bitmask.bitwidth)
    elif ident in registry.structs:
        return CStructType(ident.value)
    elif ident in registry.unions:
        return CUnionType(ident.value)
    elif ident in registry.handles:
        return CHandleType(ident)
    elif ident in registry.functions:
        fn = registry.functions[ident]
        comment = f'{fn.name}'
        return CPointerType(CTYPE_VOID, False, comment)
    else:
        raise Exception(f'unknown type: {ident}')

File: codegen-v2/src/main/kotlin/club/doki7/babel/ctype/ctype.kt
package club.doki7.babel.ctype

import club.doki7.babel.registry.ArrayType
import club.doki7.babel.registry.IdentifierType
import club.doki7.babel.registry.PointerType
import club.doki7.babel.registry.RegistryBase
import club.doki7.babel.registry.Type
import kotlin.collections.contains

sealed interface CType {
    val jType: String
    val jLayout: String
    val jLayoutType: String
    val cType: String

    val jDescriptorParamLayout: String get() = jLayout
}

data class CVoidType(override val cType: String = "void") : CType {
    override val jType: String = "void"
    override val jLayout: String get() = error("should not call `jLayout` on `void`")
    override val jLayoutType: String get() = error("should not call `jLayoutType` on `void`")
}

val voidType = CVoidType()

data class CPointerType(
    val pointee: CType,
    val const: Boolean,
    val pointerToOne: Boolean,
    val comment: String?,
) : CType {
    override val jType: String = if (comment != null) {
        """@Pointer(comment="$comment") MemorySegment"""
    }
    else {
        """@Pointer(comment="void*") MemorySegment"""
    }

    override val jLayout: String = if (pointee is CVoidType) {
        "ValueLayout.ADDRESS"
    }
    else {
        """ValueLayout.ADDRESS.withTargetLayout(${pointee.jLayout})"""
    }

    override val jLayoutType: String = "AddressLayout"

    override val cType: String = """${pointee.cType}${if (const) " const" else ""}*"""
}

data class CHandleType(val name: String) : CType {
    override val jType: String = name
    override val jLayout: String = "ValueLayout.ADDRESS"
    override val jLayoutType: String = "AddressLayout"
    override val cType: String = name
}

data class CArrayType(val element: CType, val length: String) : CType {
    override val jType: String get() = error("should not call `jRawType` on `array`")
    override val jLayout: String = "MemoryLayout.sequenceLayout($length, ${element.jLayout})"
    override val jLayoutType: String = "SequenceLayout"
    override val cType: String = "${element.cType}[$length]"

    override val jDescriptorParamLayout: String = "ValueLayout.ADDRESS.withTargetLayout(${element.jLayout})"

    val flattened: CArrayType get() {
        if (element !is CArrayType) {
            return this
        }

        val newElement = element.flattened
        val newLength = "$length * ${element.length}"
        return CArrayType(newElement.element, newLength)
    }
}

sealed interface CNonRefType : CType {
    val jTypeNoSign: String
    val jPtrType: String
    val jPtrTypeNoAnnotation: String
    val comment: String?
}

sealed interface CFixedSizeType : CNonRefType {
    val byteSize: Int
}

data class CFixedIntType(
    override val cType: String,
    override val byteSize: Int,
    val unsigned: Boolean,
    override val comment: String? = null
) : CFixedSizeType {
    override val jType: String get() = buildString {
        if (comment != null) {
            append("@NativeType(\"$comment\") ")
        }
        if (unsigned) {
            append("@Unsigned ")
        }
        append(jTypeNoSign)
    }

    override val jLayout: String get() = when (byteSize) {
        1 -> "ValueLayout.JAVA_BYTE"
        2 -> "ValueLayout.JAVA_SHORT"
        4 -> "ValueLayout.JAVA_INT"
        8 -> "ValueLayout.JAVA_LONG"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jLayoutType: String get() = when (byteSize) {
        1 -> "OfByte"
        2 -> "OfShort"
        4 -> "OfInt"
        8 -> "OfLong"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jTypeNoSign: String get() = when (byteSize) {
        1 -> "byte"
        2 -> "short"
        4 -> "int"
        8 -> "long"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jPtrType: String get() = buildString {
        if (comment != null) {
            append("@Pointer(comment=\"$comment\") ")
        }
        if (unsigned) {
            append("@Unsigned ")
        }
        append(jPtrTypeNoAnnotation)
    }

    override val jPtrTypeNoAnnotation: String get() = when (byteSize) {
        1 -> "BytePtr"
        2 -> "ShortPtr"
        4 -> "IntPtr"
        8 -> "LongPtr"
        else -> error("unsupported byte size: $byteSize")
    }
}

data class CPlatformDependentIntType(
    override val cType: String,
    override val jType: String,
    override val jLayout: String,
    override val jTypeNoSign: String,
    override val jPtrType: String,
    override val jPtrTypeNoAnnotation: String,
    override val comment: String? = null
) : CNonRefType {
    override val jLayoutType: String get() = error("should not call `jLayoutType` on `CPlatformDependentIntType`")
}

data class CFloatType(
    override val byteSize: Int,
    override val comment: String? = null
) : CFixedSizeType {
    override val jType: String get() = when (byteSize) {
        4 -> "float"
        8 -> "double"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jLayout: String get() = when (byteSize) {
        4 -> "ValueLayout.JAVA_FLOAT"
        8 -> "ValueLayout.JAVA_DOUBLE"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jLayoutType: String = when (byteSize) {
        4 -> "OfFloat"
        8 -> "OfDouble"
        else -> error("unsupported byte size: $byteSize")
    }
    override val cType: String = when (byteSize) {
        4 -> "float"
        8 -> "double"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jTypeNoSign: String get() = jType
    override val jPtrType: String get() = when (byteSize) {
        4 -> "FloatPtr"
        8 -> "DoublePtr"
        else -> error("unsupported byte size: $byteSize")
    }
    override val jPtrTypeNoAnnotation: String get() = jPtrType
}

data class CStructType(val name: String): CType {
    override val jType: String = "I$name"
    override val jLayout: String = "$name.LAYOUT"
    override val jLayoutType: String = "StructLayout"
    override val cType: String = name
}

data class CEnumType(
    val name: String,
    val bitwidth: Int? = null,
    override val comment: String? = null
): CFixedSizeType {
    override val jType: String get() = when (bitwidth) {
        null, 32 -> {
            "@EnumType($name.class) int"
        }
        64 -> {
            "@EnumType($name.class) long"
        }
        else -> {
            error("unsupported bitwidth: $bitwidth")
        }
    }

    override val byteSize: Int get() = when (bitwidth) {
        null, 32 -> 4
        64 -> 8
        else -> error("unsupported bitwidth: $bitwidth")
    }

    override val jLayout: String get() = when (bitwidth) {
        null, 32 -> "ValueLayout.JAVA_INT"
        64 -> "ValueLayout.JAVA_LONG"
        else -> error("unsupported bitwidth: $bitwidth")
    }

    override val jLayoutType: String = when (bitwidth) {
        null, 32 -> "OfInt"
        64 -> "OfLong"
        else -> error("unsupported bitwidth: $bitwidth")
    }

    override val cType: String = "enum $name"

    override val jTypeNoSign: String get() = when (bitwidth) {
        null, 32 -> "int"
        64 -> "long"
        else -> error("unsupported bitwidth: $bitwidth")
    }

    override val jPtrType: String = when (bitwidth) {
        null, 32 -> "@EnumType($name.class) IntPtr"
        64 -> "@EnumType($name.class) LongPtr"
        else -> error("unsupported bitwidth: $bitwidth")
    }

    override val jPtrTypeNoAnnotation: String = when (bitwidth) {
        null, 32 -> "IntPtr"
        64 -> "LongPtr"
        else -> error("unsupported bitwidth: $bitwidth")
    }
}

private val int8Type = CFixedIntType("int8_t", 1, false)
private val uint8Type = CFixedIntType("uint8_t", 1, true)
private val int16Type = CFixedIntType("int16_t", 2, false)
private val uint16Type = CFixedIntType("uint16_t", 2, true)
private val int32Type = CFixedIntType("int32_t", 4, false)
private val uint32Type = CFixedIntType("uint32_t", 4, true)
private val int64Type = CFixedIntType("int64_t", 8, false)
private val uint64Type = CFixedIntType("uint64_t", 8, true)
private val floatType = CFloatType(4)
private val doubleType = CFloatType(8)

private val cIntType = int32Type
private val cUIntType = uint32Type
private val cLongType = CPlatformDependentIntType(
    "long",
    "long",
    "NativeLayout.C_LONG",
    "long",
    "CLongPtr",
    "CLongPtr"
)
private val cSizeType = CPlatformDependentIntType(
    "size_t",
    "long",
    "NativeLayout.C_SIZE_T",
    "long",
    "PointerPtr",
    "PointerPtr"
)

private fun pvoidType(comment: String) = CPointerType(voidType, const = false, pointerToOne = false, comment = comment)

private val knownTypes = mapOf(
    // Fundamental types
    "void" to voidType,
    "int8_t" to int8Type,
    "uint8_t" to uint8Type,
    "int16_t" to int16Type,
    "uint16_t" to uint16Type,
    "int32_t" to int32Type,
    "uint32_t" to uint32Type,
    "int64_t" to int64Type,
    "uint64_t" to uint64Type,
    "float" to floatType,
    "double" to doubleType,
    "int" to cIntType,
    "unsigned" to cUIntType,
    "unsigned int" to cUIntType,
    "long" to cLongType,
    "size_t" to cSizeType,

    "char" to int8Type,
    "unsigned char" to uint8Type,
    "short" to int16Type,
    "unsigned short" to uint16Type,

    // GLES2 base types
    "GLbyte" to int8Type.copy(comment = "GLbyte"),
    "GLubyte" to uint8Type.copy(comment = "GLubyte"),
    "GLchar" to int8Type.copy(comment = "GLchar"),
    "GLuchar" to uint8Type.copy(comment = "GLuchar"),
    "GLclampf" to floatType.copy(comment = "GLclampf"),
    "GLfixed" to int32Type.copy(comment = "GLfixed"),
    "GLint" to int32Type.copy(comment = "GLint"),
    "GLuint" to uint32Type.copy(comment = "GLuint"),
    "GLshort" to int16Type.copy(comment = "GLshort"),
    "Glushort" to uint16Type.copy(comment = "GLushort"),
    "GLfloat" to floatType.copy(comment = "GLfloat"),
    "GLvoid" to voidType,
    "GLenum" to int32Type.copy(comment = "GLenum"),
    "GLsizei" to int32Type.copy(comment = "GLsizei"),
    "GLsizeiptr" to cSizeType.copy(comment = "GLsizeiptr"),
    "GLintptr" to cSizeType.copy(comment = "GLintptr"),
    "GLbitfield" to uint32Type.copy(comment = "GLbitfield"),
    "GLboolean" to uint8Type.copy(comment = "GLboolean"),

    // Vulkan base types
    "VkSampleMask" to uint32Type.copy(comment = "VkSampleMask"),
    "VkBool32" to uint32Type.copy(comment = "VkBool32"),
    "VkFlags" to uint32Type.copy(comment = "VkFlags"),
    "VkFlags64" to uint64Type.copy(comment = "VkFlags64"),
    "VkDeviceSize" to uint64Type.copy(comment = "VkDeviceSize"),
    "VkDeviceAddress" to uint64Type.copy(comment = "VkDeviceAddress"),
    "VkRemoteAddressNV" to pvoidType("VkRemoteAddressNV"),

    // Android
    "ANativeWindow" to pvoidType("ANativeWindow"),
    "AHardwareBuffer" to pvoidType("AHardwareBuffer"),

    // DirectFB
    "IDirectFB" to pvoidType("IDirectFB"),
    "IDirectFBSurface" to pvoidType("IDirectFBSurface"),

    // iOS or macOS
    "id" to pvoidType("id"),
    "CAMetalLayer" to pvoidType("CAMetalLayer"),
    "GgpFrameToken" to uint32Type.copy(comment = "GgpFrameToken"),
    "GgpStreamDescriptor" to uint32Type.copy(comment = "GgpStreamDescriptor"),
    "IOSurfaceRef" to pvoidType("IOSurfaceRef"),
    "MTLBuffer_id" to pvoidType("MTLBuffer_id"),
    "MTLCommandQueue_id" to pvoidType("MTLCommandQueue_id"),
    "MTLDevice_id" to pvoidType("MTLDevice_id"),
    "MTLSharedEvent_id" to pvoidType("MTLSharedEvent_id"),
    "MTLTexture_id" to pvoidType("MTLTexture_id"),
    "CGDirectDisplayID" to uint32Type.copy(comment = "CGDirectDisplayID"),

    // QNX
    "_screen_buffer" to pvoidType("_screen_buffer"),
    "_screen_context" to pvoidType("_screen_context"),
    "_screen_window" to pvoidType("_screen_window"),

    // Wayland
    "wl_display" to pvoidType("wl_display"),
    "wl_surface" to pvoidType("wl_surface"),
    "wl_output" to voidType.copy(cType = "wl_output"),

    // Windows
    "DWORD" to uint32Type.copy(comment = "DWORD"),
    "HANDLE" to pvoidType("HANDLE"),
    "HINSTANCE" to pvoidType("HINSTANCE"),
    "HMONITOR" to pvoidType("HMONITOR"),
    "HWND" to pvoidType("HWND"),
    "LPCWSTR" to CPointerType(uint16Type, const = true, pointerToOne = false, comment = "LPCWSTR"),
    "HGLRC" to pvoidType("HGLRC"),
    "SECURITY_ATTRIBUTES" to voidType,

    // X11
    "Display" to pvoidType("Display"),
    "RROutput" to cLongType.copy(comment = "RROutput"),
    "RRCrtc" to cLongType.copy(comment = "RRCrtc"),
    "VisualID" to cLongType.copy(comment = "VisualID"),
    "Window" to cLongType.copy(comment = "Window"),
    "GLXContext" to pvoidType("GLXContext"),
    "GLXWindow" to cLongType.copy(comment = "GLXWindow"),
    "xcb_connection_t" to voidType.copy(cType = "xcb_connection_t"),
    "xcb_visualid_t" to uint32Type.copy(comment = "xcb_visualid_t"),
    "xcb_window_t" to uint32Type.copy(comment = "xcb_window_t"),
    "xcb_handle_t" to uint32Type.copy(comment = "xcb_handle_t"),

    // EGL
    "EGLDisplay" to pvoidType("EGLDisplay"),
    "EGLContext" to pvoidType("EGLContext"),
    "EGLSurface" to pvoidType("EGLSurface"),

    // MESA
    "OSMesaContext" to pvoidType("OSMesaContext"),

    // NvSciBuf / NvSciSync
    "NvSciBufAttrList" to pvoidType("NvSciBufAttrList"),
    "NvSciBufObj" to pvoidType("NvSciBufObj"),
    "NvSciSyncAttrList" to pvoidType("NvSciSyncAttrList"),
    "NvSciSyncObj" to pvoidType("NvSciSyncObj"),

    // FUCHSIA
    "zx_handle_t" to uint32Type.copy(comment = "zx_handle_t"),
)

fun lowerType(registry: RegistryBase, refRegistries: List<RegistryBase>, type: Type): CType {
    return when(type) {
        is ArrayType -> {
            if (!type.length.value.isNumeric()) {
                if (!registry.constants.contains(type.length)
                    && !refRegistries.any { it.constants.contains(type.length) }) {
                    error("array type referred to an unknown constant ${type.length}")
                }
            }

            val elementType = lowerType(registry, refRegistries, type.element)
            CArrayType(elementType, type.length.value)
        }
        is PointerType -> {
            if (type.pointee is IdentifierType &&
                (registry.opaqueTypedefs.contains(type.pointee.ident)
                 || refRegistries.any { it.opaqueTypedefs.contains(type.pointee.ident) })) {
                return CHandleType(type.pointee.ident.value)
            }

            val pointee = lowerType(registry, refRegistries, type.pointee)
            CPointerType(pointee, type.const, type.pointerToOne, comment=null)
        }
        is IdentifierType -> lowerIdentifierType(registry, refRegistries, type)
    }
}

fun lowerIdentifierType(
    registry: RegistryBase,
    refRegistries: List<RegistryBase>,
    type: IdentifierType
): CType {
    val lookupResult = identifierTypeLookup(registry, refRegistries, type)
    if (lookupResult != null) {
        return lookupResult
    }

    for (refRegistry in refRegistries) {
        val lookupResult = identifierTypeLookup(refRegistry, refRegistries, type)
        if (lookupResult != null) {
            return lookupResult
        }
    }

    return if (knownTypes.containsKey(type.ident.value)) {
        knownTypes[type.ident.value]!!
    } else {
        error("unknown type ${type.ident.value}")
    }
}

fun identifierTypeLookup(registry: RegistryBase, refRegistries: List<RegistryBase>, type: IdentifierType) =
    if (registry.structures.contains(type.ident)) {
        CStructType(type.ident.value)
    }
    else if (registry.unions.contains(type.ident)) {
        CStructType(type.ident.value)
    }
    else if (registry.enumerations.contains(type.ident)) {
        CEnumType(type.ident.value)
    }
    else if (registry.bitmasks.contains(type.ident)) {
        CEnumType(type.ident.value, bitwidth=registry.bitmasks[type.ident]!!.bitwidth)
    }
    else if (registry.opaqueHandleTypedefs.contains(type.ident)) {
        CHandleType(type.ident.value)
    }
    else if (registry.functionTypedefs.contains(type.ident)) {
        CPointerType(voidType, false, pointerToOne = true, comment=type.ident.value)
    }
    else if (registry.aliases.contains(type.ident)) {
        val alias = registry.aliases[type.ident]!!
        lowerType(registry, refRegistries, alias.type)
    }
    else if (knownTypes.containsKey(type.ident.value)) {
        knownTypes[type.ident.value]!!
    } else {
        null
    }

private fun String.isNumeric() = all { it.isDigit() }
